controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +202,87 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
+func MatchesTemplateClonedFrom(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		infraObj, err := external.Get(ctx, c, &machine.Spec.InfrastructureRef, kcp.Namespace)
+		// Return true here because failing to get infrastructure machine should not be considered as unmatching.
+		if err != nil {
+			return true
+		}
+
+		clonedFromName, ok1 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromNameAnnotation]
+		clonedFromGroupKind, ok2 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
+		if !ok1 || !ok2 {
+			// All kcp cloned infra machines should have this annotation.
+			// Missing the annotation may be due to older version machines or adopted machines.
+			// Should not be considered as mismatch.
+			return true
+		}
+
+		//Check if infra machine's InfrastructureTemplate is matching the kcp InfrastructureTemplate.
+		if clonedFromName != kcp.Spec.InfrastructureTemplate.Name ||
+			clonedFromGroupKind != kcp.Spec.InfrastructureTemplate.GroupVersionKind().GroupKind().String() {
+			return false
+		}
+		return true
+	}
+}
+
+// MatchesKubernetesVersion returns a filter to find all machines that match a given Kubernetes version.
+func MatchesKubernetesVersion(kubernetesVersion string) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		return *machine.Spec.Version == kubernetesVersion
+	}
+}
+
+// MatchesKubeadmBootstrapConfig checks if machine's KubeadmConfigSpec is equivalent with KCP's KubeadmConfigSpec.
+func MatchesKubeadmBootstrapConfig(ctx context.Context, c client.Client, kcpConfigSpec bootstrapv1.KubeadmConfigSpec) Func {
+	return func(machine *clusterv1.Machine) bool {
+		bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+		if bootstrapRef == nil {
+			// Missing bootstrap reference should not be considered as unmatching.
+			return true
+		}
+
+		bootstrapObj := &bootstrapv1.KubeadmConfig{}
+		if err := c.Get(ctx, client.ObjectKey{Name: bootstrapRef.Name, Namespace: machine.Namespace}, bootstrapObj); err != nil {
+			// Return true here because failing to get KubeadmConfig should not be considered as unmatching.
+			return true
+		}
+
+		// TODO: do we need this check?
+		if bootstrapObj.Spec.InitConfiguration == nil && bootstrapObj.Spec.JoinConfiguration == nil && bootstrapObj.Spec.ClusterConfiguration == nil {
+			return false
+		}
+		kcpConfigSpecLocal := kcpConfigSpec.DeepCopy()
+		if bootstrapObj.Spec.InitConfiguration == nil {
+			kcpConfigSpecLocal.InitConfiguration = nil
+		}
+		if bootstrapObj.Spec.JoinConfiguration == nil {
+			kcpConfigSpecLocal.JoinConfiguration = nil
+		}
+		if bootstrapObj.Spec.ClusterConfiguration == nil {
+			kcpConfigSpecLocal.ClusterConfiguration = nil
+		}
+
+		return reflect.DeepEqual(bootstrapObj.Spec, *kcpConfigSpecLocal)
```

Comment:
I don't think we can necessarily rely on a DeepEqual here, since we apply a certain level of defaulting for KubeadmConfig (at least in the case of JoinConfig == nil, you might need to leverage SemanticMerge from bootstrap/kubeadm/api/equality/semantic.go or similar when doing the comparison.

Reply:
@ncdc originally suggested to use DeepEqual here

Reply:
We still need semantic merge for the fields such as ControlPlaneEndpoint in ClusterConfiguration ( Machine's KubeadmConfig has this but KCP may not).